File: packages/cpp/src/explore/explorer.cpp
#include <handoffkit/explore/explorer.hpp>

#include <queue>
#include <set>
#include <utility>

namespace handoffkit {
namespace explore {

WebExplorer::WebExplorer(TransportPtr transport) : transport_(std::move(transport)) {
    if (!transport_) transport_ = std::make_shared<MapTransport>();
}

void WebExplorer::set_transport(TransportPtr transport) {
    transport_ = transport ? std::move(transport) : std::make_shared<MapTransport>();
}

void WebExplorer::set_policy(ExplorePolicy policy) { policy_ = std::move(policy); }

TransportRequest WebExplorer::make_request(std::string_view url, const ExplorePolicy& policy) const {
    TransportRequest req;
    req.url = std::string(url);
    req.timeout_ms = policy.timeout_ms;
    req.max_body_bytes = policy.max_body_bytes;
    req.follow_redirects = policy.follow_redirects;
    req.max_redirects = policy.max_redirects;
    req.headers["User-Agent"] = policy.user_agent;
    req.headers["Accept"] = "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8";
    for (const auto& [k, v] : policy.extra_headers) {
        req.headers[k] = v;
    }
    return req;
}

Result<ExploreStep> WebExplorer::fetch_one(std::string_view url, int depth,
                                           const ExplorePolicy& policy) const {
    ExploreStep step;
    step.depth = depth;
    step.url = std::string(url);

    auto parts = parse_url(url);
    if (!parts.valid || parts.host.empty()) {
        step.success = false;
        step.error = "invalid url";
        return Result<ExploreStep>::success(std::move(step));
    }
    if (!host_allowed(parts.host, policy)) {
        step.success = false;
        step.error = "host denied by policy: " + parts.host;
        return Result<ExploreStep>::success(std::move(step));
    }
    if (!transport_) {
        step.success = false;
        step.error = "no transport configured";
        return Result<ExploreStep>::success(std::move(step));
    }

    auto resp = transport_->get(make_request(url, policy));
    step.status = resp.status;
    step.final_url = resp.final_url.empty() ? step.url : resp.final_url;
    step.raw_body_bytes = resp.body.size();
    if (!resp.error.empty()) {
        step.success = false;
        step.error = resp.error;
        return Result<ExploreStep>::success(std::move(step));
    }
    if (resp.status < 200 || resp.status >= 400) {
        step.success = false;
        step.error = "HTTP status " + std::to_string(resp.status);
        // still try extract if body present
    }
    auto page = extract_page(step.final_url, resp.body, policy);
    step.title = std::move(page.title);
    step.text = std::move(page.text);
    step.markdown = std::move(page.markdown);
    step.links = std::move(page.links);

    // classify blocked outbound links
    const std::string origin = normalize_host(parts.host);
    for (const auto& link : step.links) {
        if (link.absolute.empty()) continue;
        if (!url_allowed(link.absolute, policy, origin)) {
            step.blocked_links.push_back(link.absolute);
        }
    }
    if (step.error.empty()) step.success = true;
    return Result<ExploreStep>::success(std::move(step));
}

Result<ExploreResult> WebExplorer::fetch(std::string_view url) const {
    return fetch(url, policy_);
}

Result<ExploreResult> WebExplorer::fetch(std::string_view url, const ExplorePolicy& policy) const {
    ExploreResult result;
    result.start_url = std::string(url);
    result.policy = policy;
    result.metadata["transport"] = transport_ ? transport_->name() : "none";
    result.metadata["mode"] = "fetch";

    auto step_r = fetch_one(url, 0, policy);
    if (!step_r) return step_r.error();
    auto step = step_r.value();
    step.step_index = 0;
    result.steps.push_back(step);
    if (step.success) {
        result.success = true;
        result.pages_fetched = 1;
        result.max_depth_reached = 0;
        result.final_url = step.final_url;
        result.title = step.title;
        result.text = step.text;
        result.markdown = step.markdown;
        result.links = step.links;
    } else {
        result.success = false;
        result.error = step.error;
        result.final_url = step.final_url;
    }
    return result;
}

Result<ExploreResult> WebExplorer::explore(std::string_view start_url) const {
    return explore(start_url, policy_);
}

Result<ExploreResult> WebExplorer::explore(std::string_view start_url, const ExplorePolicy& policy) const {
    ExploreResult result;
    result.start_url = std::string(start_url);
    result.policy = policy;
    result.metadata["transport"] = transport_ ? transport_->name() : "none";
    result.metadata["mode"] = "explore";

    auto start_parts = parse_url(start_url);
    if (!start_parts.valid || start_parts.host.empty()) {
        result.success = false;
        result.error = "invalid start_url";
        return result;
    }
    if (!host_allowed(start_parts.host, policy)) {
        result.success = false;
        result.error = "start host denied by policy: " + start_parts.host;
        return result;
    }

    const std::string origin_host = normalize_host(start_parts.host);
    std::set<std::string> visited;
    // BFS: (url, depth)
    std::queue<std::pair<std::string, int>> q;
    q.push({std::string(start_url), 0});

    int step_index = 0;
    while (!q.empty() && result.pages_fetched < policy.max_pages) {
        auto [url, depth] = q.front();
        q.pop();
        if (visited.count(url)) continue;
        if (depth > policy.max_depth) continue;
        visited.insert(url);

        auto step_r = fetch_one(url, depth, policy);
        if (!step_r) return step_r.error();
        auto step = step_r.value();
        step.step_index = step_index++;
        result.steps.push_back(step);
        result.max_depth_reached = std::max(result.max_depth_reached, depth);

        if (!step.success) {
            // count failed attempts toward page budget only if we got a response body attempt
            // still continue exploring other queued URLs
            continue;
        }
        ++result.pages_fetched;
        if (result.final_url.empty()) {
            result.final_url = step.final_url;
            result.title = step.title;
            result.text = step.text;
            result.markdown = step.markdown;
        } else {
            // append text / markdown excerpts for multi-page
            if (!step.text.empty() && static_cast<int>(result.text.size()) < policy.max_text_chars) {
                if (!result.text.empty()) result.text += "\n\n";
                result.text += step.text;
                if (static_cast<int>(result.text.size()) > policy.max_text_chars) {
                    result.text.resize(static_cast<std::size_t>(policy.max_text_chars));
                    result.text += "...[truncated]";
                }
            }
            if (!step.markdown.empty()) {
                const int md_cap = policy.max_markdown_chars > 0 ? policy.max_markdown_chars
                                                                : policy.max_text_chars;
                if (static_cast<int>(result.markdown.size()) < md_cap) {
                    if (!result.markdown.empty()) result.markdown += "\n\n---\n\n";
                    result.markdown += step.markdown;
                    if (static_cast<int>(result.markdown.size()) > md_cap) {
                        result.markdown.resize(static_cast<std::size_t>(md_cap));
                        result.markdown += "\n\n...[truncated]\n";
                    }
                }
            }
        }
        for (const auto& link : step.links) {
            result.links.push_back(link);
        }

        if (depth >= policy.max_depth) continue;
        if (result.pages_fetched >= policy.max_pages) break;

        for (const auto& link : step.links) {
            if (link.absolute.empty()) continue;
            if (visited.count(link.absolute)) continue;
            if (!url_allowed(link.absolute, policy, origin_host)) continue;
            if (static_cast<int>(visited.size()) + static_cast<int>(q.size()) > policy.max_pages * 4) {
                break;  // avoid unbounded queue
            }
            q.push({link.absolute, depth + 1});
        }
    }

    result.success = result.pages_fetched > 0;
    if (!result.success && result.error.empty()) {
        if (!result.steps.empty()) {
            result.error = result.steps.front().error.empty()
                ? "no pages fetched successfully"
                : result.steps.front().error;
        } else {
            result.error = "explore produced no steps";
        }
    }
    result.metadata["visited"] = static_cast<int>(visited.size());
    return result;
}

Result<ExploreResult> web_fetch(std::string_view url, TransportPtr transport, ExplorePolicy policy) {
    WebExplorer ex(std::move(transport));
    return ex.fetch(url, policy);
}

Result<ExploreResult> web_explore(std::string_view start_url, TransportPtr transport, ExplorePolicy policy) {
    WebExplorer ex(std::move(transport));
    return ex.explore(start_url, policy);
}

}  // namespace explore
}  // namespace handoffkit

File: js/packages/browser/src/explorer.js
import {
  ExplorePolicy,
  hostAllowed,
  normalizeHost,
  parseUrl,
  resultToDict,
  urlAllowed,
} from "./types.js";
import { extractPage } from "./html_extract.js";
import { MapTransport, TransportRequest } from "./transport.js";
import { detectSoftBlock } from "./util.js";

export class WebExplorer {
  constructor(transport = null, policy = null) {
    this.transport = transport ?? new MapTransport();
    this.policy = policy instanceof ExplorePolicy ? policy : new ExplorePolicy(policy ?? {});
  }

  setTransport(transport) {
    this.transport = transport ?? new MapTransport();
    return this;
  }

  setPolicy(policy) {
    this.policy = policy instanceof ExplorePolicy ? policy : new ExplorePolicy(policy ?? {});
    return this;
  }

  _makeRequest(url, policy) {
    return new TransportRequest({
      url,
      timeoutMs: policy.timeoutMs,
      maxBodyBytes: policy.maxBodyBytes,
      followRedirects: policy.followRedirects,
      maxRedirects: policy.maxRedirects,
      headers: {
        "User-Agent": policy.userAgent,
        Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        ...policy.extraHeaders,
      },
    });
  }

  async _fetchOne(url, depth, policy) {
    const step = {
      stepIndex: 0,
      depth,
      url: String(url),
      finalUrl: "",
      status: 0,
      success: false,
      error: "",
      title: "",
      text: "",
      markdown: "",
      links: [],
      rawBodyBytes: 0,
      blockedLinks: [],
    };

    const parts = parseUrl(url);
    if (!parts.valid || !parts.host) {
      step.error = "invalid url";
      return step;
    }
    if (!hostAllowed(parts.host, policy)) {
      step.error = `host denied by policy: ${parts.host}`;
      return step;
    }
    if (!this.transport) {
      step.error = "no transport configured";
      return step;
    }

    const resp = await this.transport.get(this._makeRequest(url, policy));
    step.status = resp.status;
    step.finalUrl = resp.finalUrl || step.url;
    step.rawBodyBytes = resp.body?.length ?? 0;
    if (resp.error) {
      step.error = resp.error;
      return step;
    }
    const soft = detectSoftBlock(resp.body ?? "", resp.status);
    if (soft.blocked) {
      step.error = soft.reason;
      step.success = false;
      return step;
    }
    if (resp.status < 200 || resp.status >= 400) {
      step.error = `HTTP status ${resp.status}`;
    }
    const page = extractPage(step.finalUrl, resp.body ?? "", policy);
    step.title = page.title;
    step.text = page.text;
    step.markdown = page.markdown;
    step.links = page.links;

    const origin = normalizeHost(parts.host);
    for (const link of step.links) {
      if (!link.absolute) continue;
      if (!urlAllowed(link.absolute, policy, origin)) {
        step.blockedLinks.push(link.absolute);
      }
    }
    if (!step.error) step.success = true;
    return step;
  }

  async fetch(url, policyOverride = null) {
    const policy =
      policyOverride instanceof ExplorePolicy
        ? policyOverride
        : policyOverride
          ? new ExplorePolicy(policyOverride)
          : this.policy;

    const result = {
      success: false,
      startUrl: String(url),
      finalUrl: "",
      pagesFetched: 0,
      maxDepthReached: 0,
      title: "",
      text: "",
      markdown: "",
      links: [],
      steps: [],
      policy,
      error: "",
      metadata: {
        transport: this.transport?.name?.() ?? "none",
        mode: "fetch",
      },
    };

    const step = await this._fetchOne(url, 0, policy);
    step.stepIndex = 0;
    result.steps.push(step);
    if (step.success) {
      result.success = true;
      result.pagesFetched = 1;
      result.finalUrl = step.finalUrl;
      result.title = step.title;
      result.text = step.text;
      result.markdown = step.markdown;
      result.links = step.links;
    } else {
      result.error = step.error;
      result.finalUrl = step.finalUrl;
    }
    return result;
  }

  async explore(startUrl, policyOverride = null) {
    const policy =
      policyOverride instanceof ExplorePolicy
        ? policyOverride
        : policyOverride
          ? new ExplorePolicy(policyOverride)
          : this.policy;

    const result = {
      success: false,
      startUrl: String(startUrl),
      finalUrl: "",
      pagesFetched: 0,
      maxDepthReached: 0,
      title: "",
      text: "",
      markdown: "",
      links: [],
      steps: [],
      policy,
      error: "",
      metadata: {
        transport: this.transport?.name?.() ?? "none",
        mode: "explore",
      },
    };

    const startParts = parseUrl(startUrl);
    if (!startParts.valid || !startParts.host) {
      result.error = "invalid start_url";
      return result;
    }
    if (!hostAllowed(startParts.host, policy)) {
      result.error = `start host denied by policy: ${startParts.host}`;
      return result;
    }

    const originHost = normalizeHost(startParts.host);
    const visited = new Set();
    /** @type {Array<{ url: string, depth: number }>} */
    const queue = [{ url: String(startUrl), depth: 0 }];
    let stepIndex = 0;

    while (queue.length && result.pagesFetched < policy.maxPages) {
      const { url, depth } = queue.shift();
      if (visited.has(url)) continue;
      if (depth > policy.maxDepth) continue;
      visited.add(url);

      const step = await this._fetchOne(url, depth, policy);
      step.stepIndex = stepIndex++;
      result.steps.push(step);
      result.maxDepthReached = Math.max(result.maxDepthReached, depth);

      if (!step.success) continue;

      result.pagesFetched += 1;
      if (!result.finalUrl) {
        result.finalUrl = step.finalUrl;
        result.title = step.title;
        result.text = step.text;
        result.markdown = step.markdown;
      } else {
        if (step.text && result.text.length < policy.maxTextChars) {
          if (result.text) result.text += "\n\n";
          result.text += step.text;
          if (result.text.length > policy.maxTextChars) {
            result.text = `${result.text.slice(0, policy.maxTextChars)}...[truncated]`;
          }
        }
        if (step.markdown) {
          const mdCap = policy.maxMarkdownChars > 0 ? policy.maxMarkdownChars : policy.maxTextChars;
          if (result.markdown.length < mdCap) {
            if (result.markdown) result.markdown += "\n\n---\n\n";
            result.markdown += step.markdown;
            if (result.markdown.length > mdCap) {
              result.markdown = `${result.markdown.slice(0, mdCap)}\n\n...[truncated]\n`;
            }
          }
        }
      }
      result.links.push(...step.links);

      if (depth >= policy.maxDepth) continue;
      if (result.pagesFetched >= policy.maxPages) break;

      for (const link of step.links) {
        if (!link.absolute || visited.has(link.absolute)) continue;
        if (!urlAllowed(link.absolute, policy, originHost)) continue;
        if (visited.size + queue.length > policy.maxPages * 4) break;
        queue.push({ url: link.absolute, depth: depth + 1 });
      }
    }

    result.success = result.pagesFetched > 0;
    if (!result.success && !result.error) {
      if (result.steps.length) {
        result.error = result.steps[0].error || "no pages fetched successfully";
      } else {
        result.error = "explore produced no steps";
      }
    }
    result.metadata.visited = visited.size;
    return result;
  }
}

export async function webFetch(url, transport = null, policy = {}) {
  const explorer = new WebExplorer(transport);
  return explorer.fetch(url, policy);
}

export async function webExplore(startUrl, transport = null, policy = {}) {
  const explorer = new WebExplorer(transport);
  return explorer.explore(startUrl, policy);
}

export { resultToDict };
